### PR TITLE
Convert protocol-relative URLs in the string resolver

### PR DESCRIPTION
**Pull request by aschempp**
The Symfony `UrlHelper::getAbsoluteUrl` ignores protocol relative URLs, but our `ContentUrlResult` does not.

### CODE REVIEW

File: core-bundle/src/Routing/Content/StringResolver.php [hunk 1]
```diff
@@ -36,6 +38,12 @@ public function resolve(object $content): ContentUrlResult|null
             $url = $this->urlHelper->getAbsoluteUrl($url);
         }
 
+        // Resolve protocol-relative URLs that are ignored by the UrlHelper
+        if (str_starts_with($url, '//')) {
+            $protocol = $this->requestStack->getCurrentRequest()?->getScheme() ?? 'http';
```
Comment (ausi) [inline core-bundle/src/Routing/Content/StringResolver.php]:
```suggestion
            $protocol = $this->requestStack->getCurrentRequest()?->getScheme() ?? 'https';
```
Shouldn’t the default be HTTPS?

Comment (aschempp) [reply core-bundle/src/Routing/Content/StringResolver.php]: I don't think we can assume that?

Comment (leofeyer) [reply core-bundle/src/Routing/Content/StringResolver.php]: I think we can.

Comment (ausi) [reply core-bundle/src/Routing/Content/StringResolver.php]: As HTTPS is also the default in most browsers nowadays, I think we should default to HTTPS everywhere we can.

Comment (aschempp) [reply core-bundle/src/Routing/Content/StringResolver.php]: Browsers seem to use `https` as default (see https://blog.chromium.org/2021/03/a-safer-default-for-navigation-https.html), but they fall back to `http`. They would not fall back in our case, right?

Comment (ausi) [reply core-bundle/src/Routing/Content/StringResolver.php]: > They would not fall back in our case, right?

For explicit URLs (not a domain entered by the user) they do not fallback one way or the other (with some exceptions like HSTS).

In practice this fallback will never be used anyways. I just thought that we should generally fallback to `https` for security reasons.

Maybe the following would be a better choice here?

```suggestion
            $protocol = $this->requestStack->getCurrentRequest()?->getScheme() ?? throw new RuntimeException('Unable to generate absolute URL because of missing request scheme.');
```

Comment (aschempp) [reply core-bundle/src/Routing/Content/StringResolver.php]: I don't understand. ANY website supports http, MOST websites support https and redirect to https if you access it on http. Why would it be better to error or use a protocol that is possibly not supported, instead of using the one we KNOW is supported. It _might_ cause a redirect, but **it works!**?? 🤷

Comment (ausi) [reply core-bundle/src/Routing/Content/StringResolver.php]: > Why would it be better to error…

Because a URL starting with `//` cannot be resolved if there is no current request. If this happens, there is a bug somewhere that needs to be fixed.

> …or use a protocol that is possibly not supported…

To avoid using a protocol that is guaranteed to be dangerous. Using `http` as a fallback here would be a prefect opportunity for an attacker as they just need to somehow manage to inject `//`

> It _might_ cause a redirect, but **it works!**?? 🤷

Works for the attacke**r!**?? 🤷

Comment (aschempp) [reply core-bundle/src/Routing/Content/StringResolver.php]: well I disagree 🤷

Comment (leofeyer) [reply core-bundle/src/Routing/Content/StringResolver.php]: Let‘s just throw the exception in this case, then we don‘t have to guess a protocol.

File: core-bundle/src/Routing/Content/ContentUrlResult.php [hunk 2]
```diff
@@ -14,15 +14,14 @@
...
-use Nyholm\Psr7\Uri;
 
 final class ContentUrlResult
 {
     private bool $redirect = false;
 
     public function __construct(public readonly object|string $content)
     {
-        if (\is_string($content) && !(new Uri($content))->getScheme()) {
+        if (\is_string($content) && !parse_url($content, PHP_URL_SCHEME)) {
```
Comment (leofeyer) [inline core-bundle/src/Routing/Content/ContentUrlResult.php:24]: Is this a related change?

**Comment by leofeyer** [comment]
Tests are still failing, but otherwise 👍.

File: core-bundle/src/Routing/Content/ContentUrlResult.php (hunk 2 above)
Comment (aschempp) [reply core-bundle/src/Routing/Content/ContentUrlResult.php:24]: well I just stumbeled upon this, it's a dependency-free way to do the same.

**Comment by aschempp** [comment]
tests fixed (and extended) 😎

File: core-bundle/src/Routing/Content/ContentUrlResult.php (hunk 2 above)
Comment (leofeyer) [reply core-bundle/src/Routing/Content/ContentUrlResult.php:24]: Yes, but `Nyholm\Psr7\Uri` is a complete PSR-7 URI implementation that does more than `parse_url()`. It normalizes URIs to lower case for example. So to stay consistent with the rest of the code base, we should continue to use it.

Comment (ausi) [reply core-bundle/src/Routing/Content/ContentUrlResult.php:24]: As we already use this in other places like https://github.com/contao/contao/blob/98974328a16e96ebe1b0e789cd256a5f0c8150c0/core-bundle/src/Routing/Content/StringResolver.php#L35 I think it is OK to use this here. To check if there is a schema at all, the two variants are functionally equivalent, but `parse_url()` should be way faster. _(Not sure if speed is relevant for `ContentUrlResult`?)_

**Review by ausi — APPROVED** [review]

File: core-bundle/src/Routing/Content/ContentUrlResult.php (hunk 2 above)
Comment (leofeyer) [reply core-bundle/src/Routing/Content/ContentUrlResult.php:24]: Do you mean in general or just for the `PHP_URL_SCHEME` case? Because we indeed use `parse_url()` in several other places as well, even to get the hostname:

https://github.com/contao/contao/blob/591a469b595e17a9cf3458cb8b1e54f264228254/news-bundle/src/Migration/FeedMigration.php#L110

I would then check the other uses of `Nyholm\Psr7\Uri`, too.

Comment (ausi) [reply core-bundle/src/Routing/Content/ContentUrlResult.php:24]: `PHP_URL_HOST` should be fine too, if your use case is also case-insensitive. 
But for port, path, query and fragment we should not use `parse_url()`.

**Comment by leofeyer** [comment]
Thank you @aschempp.

File: core-bundle/src/Routing/Content/ContentUrlResult.php (hunk 2 above)
Comment (leofeyer) [reply core-bundle/src/Routing/Content/ContentUrlResult.php:24]: I would prefer to consistently use the library, but let‘s discuss this in #6984.
